text/0000-dockerfiles.md

```diff
@@ -0,0 +1,113 @@
+# Meta
+[meta]: #meta
+- Name: Support Dockerfiles
+- Start Date: 2021-06-30
+- Author(s): sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167)
+- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)
+
+# Summary
+[summary]: #summary
+
+This RFC introduces functionality for customizing base images, as an alternative to stackpacks.
+
+# Motivation
+[motivation]: #motivation
+
+Relying on Dockerfiles for base image generation and manipulation allows us to apply buildpacks only to the problem that they solve best: managing application runtimes and dependencies.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we replace stackpacks with multi-purpose build-time and runtime Dockerfiles.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Note: kaniko, buildah, BuildKit, the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 
+
+### App-specified Dockerfiles
+
+A buildpack app may have a build.Dockerfile and/or run.Dockerfile in its app directory. A run.Dockerfile is applied to the selected runtime base image after the detection phase. A build.Dockerfile is applied to the build-time base image before the detection phase.
+
+Both Dockerfiles must accept `base_image` and `build_id` args. The `base_image` arg allows the lifecycle to specify the original base image. The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`.  
```

Comment:
Will it be possible to override the `base_image` ARG using the `pack build` command or any other `docker | podman | buildah build` command ?
example:

```
pack build -a build_base_image=docker.io/swnodrop/buildpack-jdk:1.1 .... -a run_base_image=....`
```

Reply:
Yes, but not directly. You can override the build-time base image by selecting a different builder (which is important, because the buildpack has to be present on the image). You can override the run image by selecting a different run image (`pack build --run-image`). This has the same outcome.

Reply:
If I follow you here, that means that the `base_image` should be hard coded within the `builder.toml | yaml` file and that we will pickup for the build a different builder ?

Reply:
The builder is always used as the build-time base image. It may be extended by a build.Dockerfile, but it cannot be replaced (given that it contains the buildpacks and lifecycle).

Reply:
So, how will it be possible to extend the builder definition to specify a `build.dockerfile` ? Will then this `build.Dockerfile` be used as first step to build/push a new image ?

Reply:
The builder would not specify a build.Dockerfile. Instead, a build.Dockerfile could be applied to the builder before a build (to generate a new builder) or during a build.